cmd/kwild/server/migration.go

```diff
@@ -0,0 +1,161 @@
+	m := &migrationClient{
+		listenAddress:    kwildCfg.MigrationConfig.MigrateFrom,
+		clt:              clt,
+		kwildCfg:         kwildCfg,
+		genesisCfg:       genesisCfg,
+		snapshotFileName: filepath.Join(kwildCfg.RootDir, "snapshot.sql.gz"),
+	}
+
+	// poll for the genesis state
+	if err = m.pollForGenesisState(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	return m.kwildCfg, m.genesisCfg, nil
+}
+
+// pollForGenesisState polls for the genesis state from the old chain
+// and downloads the genesis state to the snapshot file abnd updates the genesis config
+// and the kwild config with the configuration required for migration
+func (m *migrationClient) pollForGenesisState(ctx context.Context) error {
+	// Poll for the genesis state from the old chain
+	delay := defaultPollFrequency
+	ticker := time.NewTicker(delay)
+	defer ticker.Stop()
+
+	fmt.Println("Polling for genesis state from the old chain")
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			// Get the genesis state from the old chain
+			metadata, err := m.clt.GenesisState(ctx)
+			if err != nil {
+				continue
+			}
+
+			// Check if the genesis state is ready
+			if metadata.MigrationState.Status == types.NoActiveMigration || metadata.MigrationState.Status == types.MigrationNotStarted {
+				continue
+			}
+
+			// Genesis state should ready
+			if metadata.SnapshotMetadata == nil || metadata.GenesisConfig == nil {
+				continue
+			}
+
+			fmt.Println("Genesis state is available for download")
+			if err = m.downloadGenesisState(ctx, metadata); err != nil {
+				return err
+			}
+
+			return nil
+		}
+	}
+}
+
+func (m *migrationClient) downloadGenesisState(ctx context.Context, metadata *types.MigrationMetadata) error {
+	var genCfg chain.GenesisConfig
+	if err := json.Unmarshal(metadata.GenesisConfig, &genCfg); err != nil {
+		return err
+	}
+
+	// Save the genesis state
+	var snapshotMetadata statesync.Snapshot
+	if err := json.Unmarshal(metadata.SnapshotMetadata, &snapshotMetadata); err != nil {
+		return err // should we continue polling?
+	}
+
+	// create snapshot file
+	genesisStateFile, err := os.Create(m.snapshotFileName)
+	if err != nil {
+		return err
+	}
+
+	// retrieve all the snapshot chunks
+	for i := uint32(0); i < snapshotMetadata.ChunkCount; i++ {
+		chunk, err := m.clt.GenesisSnapshotChunk(ctx, snapshotMetadata.Height, i)
+		if err != nil {
+			return err
+		}
+		n, err := genesisStateFile.Write(chunk)
+		if err != nil {
+			return err
+		}
+		if n != len(chunk) {
+			return err
+		}
+	}
+
+	// Update the genesis config
+	m.genesisCfg.DataAppHash = genCfg.DataAppHash
+	m.genesisCfg.Validators = genCfg.Validators
+	m.genesisCfg.ConsensusParams.Migration = genCfg.ConsensusParams.Migration
+
+	// Update the kwild config
+	m.kwildCfg.AppConfig.GenesisState = m.snapshotFileName
+	// Listener extension
+	extensions := m.kwildCfg.AppConfig.Extensions
+	extensions["migrations"] = map[string]string{
```

Comment:
Why do we need to configure this here? Two concerns:

1. Isn't the start and end height only used in the genesis.json?
2. Why do we need to have `extensions["migrations"].listen_address` if we already have `cfg.migration.migrate_from`? Seems totally redundant

Reply:
Using control F, it looks like we literally do not reference these values anywhere.